docs/src/constraints.md

```diff
@@ -1,5 +1,514 @@
+creating a constraint with the same name as a variable or an existing constraint
+will result in an error. To overcome this limitation, it is possible to create
+anonymous constraints, just like it is possible to create
+[Anonymous JuMP variables](@ref). This is done by dropping the second argument
+to [`@constraint`](@ref):
+```jldoctest con1
+julia> con = @constraint(model, 2x <= 1)
+2 x <= 1.0
+```
+
+It is also possible use different comparison operators (e.g., `>=` and `==`) to
+create the following types of constraints:
+```jldoctest con1
+julia> @constraint(model, 2x >= 1)
+2 x >= 1.0
+
+julia> @constraint(model, 2x == 1)
+2 x = 1.0
+
+julia> @constraint(model, 1 <= 2x <= 3)
+2 x ∈ [1.0, 3.0]
+```
+
+Note that JuMP normalizes the constraints given by the user by moving all of the
+terms containing variables to the left-hand side, and all of the constant terms
+to the right-hand side. Thus, we get:
+```jldoctest; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, 2x + 1 <= 4x + 4)
+-2 x <= 3.0
+```
+
+## [Duality](@id constraint_duality)
+
+JuMP adopts the notion of duality from MathOptInterface. Roughly speaking, the
+dual variable associated with a constraint can be viewed as the decrease in the
+value of the objective given an infinitesimal relaxation of the constraint. For
+example, in a linear constraint, this relaxation is equivalent to `2x <= 1 + δ`
+and `2x >= 1 - δ`. If the constraint is an equality constraint, it depends on
+which direction is binding.
+
+Notably, this definition is independent of the objective sense. That is, the
+dual associated with a constraint is the same when the objective is a
+maximization and when it is a minimization. **This is different to traditional
+duality in linear programming.**
+
+The dual value associated with a constraint can be accessed using the
+[`JuMP.dual`](@ref) function. For example:
+
+```jldoctest
+julia> model = Model()
+A JuMP Model
+
+julia> @variable(model, x)
+x
+
+julia> @constraint(model, con, x <= 1)
+con : x <= 1.0
+```
+
+```@meta
+DocTestSetup = quote
+    using JuMP
+    const MOI = JuMP.MathOptInterface
+    model = Model(
+        with_optimizer(
+            MOI.Utilities.MockOptimizer,
+            JuMP.JuMPMOIModel{Float64}(),
+            eval_objective_value = false,
+            eval_variable_constraint_dual = false));
+    @variable(model, x);
+    @constraint(model, con, x <= 1);
+    @objective(model, Max, -2x);
+    JuMP.optimize!(model);
+    mock = JuMP.caching_optimizer(model).optimizer;
+    MOI.set(mock, MOI.ConstraintDual(), JuMP.optimizer_index(con), -2.0)
+end
+```
+
+```jldoctest con_duality
+julia> @objective(model, Min, -2x)
+-2 x
+
+julia> JuMP.optimize!(model)
+
+julia> JuMP.dual(con)
+-2.0
+
+julia> @objective(model, Max, 2x)
+2 x
+
+julia> JuMP.optimize!(model)
+
+julia> JuMP.dual(con)
+-2.0
+```
+
+To help users who may be less familiar with conic duality, JuMP provides the
```

Comment:
Conic duality hasn't been mentioned anywhere up to this point (although it should be)